### ISSUE TITLE
`jl_critical_error` deadlocks on `malloc`/`realloc` error

### ISSUE THREAD

**Issue by vchuravy** [JuliaLang/julia#28908]
While debugging a GC error, Kiran and I encountered the scenario that `realloc` throws internal error which triggers `jl_critical_error`, which does another malloc, which then deadlocks internally.

The error we saw from `realloc` was `malloc_consolidate(): invalid chunk size`.

**Comment by yuyichao** [comment]
Yes, jl_critical_error is known to be not safe. I deal with this when I have rr attached by just disabling it and I believe that's the best solution.

It'll be much worse if we don't try to print the backtrace at all since the users won't be able to provide any useful information when they observed a crash. I also don't think there's a way to detect if the assertion/crash is in a critical part of libc and looking up debug symbols is also next to impossible to be made safe, given that not even the backtrace collection has that guarantee.

**Comment by vtjnash** [comment]
We could set up a deadman's timer to sig-9 the process if it takes too long to finish, but I'm not much convinced that would be a good idea.

backtrace collection is documented to have this guarantee—although I think only x86/64 linux honors that (ARM, Darwin, and 32-bit Win32 are known to fail this).

**Comment by vchuravy** [comment]
Would it make sense to generally disable `jl_critical_error` when we detect that we are running under rr? e.g. on startup check the `RUNNING_UNDER_RR=1` and set a static flag that turns `jl_critical_error` into noop?

**Comment by vtjnash** [comment]
FWIW, looks like libunwind might have gotten better at this recently (https://github.com/libunwind/libunwind/pull/72)
